hail/src/main/scala/is/hail/rvd/OrderedRVD.scala

```diff
@@ -1241,6 +1241,23 @@ object OrderedRVD {
       })
   }
 
+  def alignAndZipNPartitions(
```

Comment:
I don't think there's any need for this. `alignAndZipPartitions` is one of the most subtle pieces of the join machinery, which was needed to handle cases where a row from the right RDD might need to be used in multiple partitions of the result, because of producting. That issue doesn't come up in a zip join, and the binary `orderedZipJoin` doesn't use it. It does what you've done—repartition each input to a common partitioner—then uses `czipPartitions`. You should be able to do the same, bypassing `alignAndZipNPartitions` and calling `czipNPartitions` directly.

Reply:
That looks right. You could also make an n-ary version of one or more of the existing `zipPartitions` methods on `OrderedRVD`, to make it a bit more encapsulated.

> I'm inclined to leave it in in case we decide to support true multi way joins in the future.

It's up to you and @catoverdrive, but I'd be a bit concerned about committing unused and untested code to master. I'm also not convinced we would ever have need of an `alignAndZipNPartitions`. I was wrong before about the issue being products—the core issue is handling cases where the left rvd has a longer key than what is being joined on, and wanting to keep that longer key in the result. The only n-ary generalization I can think of is where the first rvd has a longer key than the rest, where the rest all have the same key. It's hard for me to imagine a use case that would require that. All multi-way joins with equal keys can be handled by a simple zipNPartitions approach.

Reply:
@patrick-schultz if that's the case, I think we should remove it for now until we need it. (This version doesn't handle the case where the keys are of different lengths)